Jenkinsfile

```diff
@@ -561,7 +561,7 @@ pipeline {
                             dir 'utils/docker'
                             label 'docker_runner'
                             additionalBuildArgs dockerBuildArgs(add_repos:false) +
-                                                " --pull --build-arg CB0=" + current_time.get(Calendar.WEEK_OF_YEAR)
+                                                " --build-arg CB0=" + current_time.get(Calendar.WEEK_OF_YEAR)
```

Comment:
You still need the `--pull` to check our Nexus DockerHub proxy for updates to the base image don't you?  Otherwise we are back to never checking (or Nexus DockerHub proxy now) if the base docker image is newer aren't we?

All that proxying on Nexus does is shields DockerHub from all of our `--pull` hits.  Well, as well as theoretically improving rebuild performance by making the pull local.

What is unclear at this point is how we point our docker builds at our proxy instead of defaulting to DockerHub.  I guess we need to provide an argument of some sort pointing at Nexus.  I would think that needs to be here in this `additionalBuildArgs`.

Reply:
As I said @JohnMalmberg is working on this.  The hope is we can have a weekly "docker --pull" job which will refresh everything once a week, and then all images will use the same base (updating the contents of course).  Whilst that may not be in place yet the correct thing for this PR is to assume that fedora:latest is correct and just use it.

Reply:
> As I said @JohnMalmberg is working on this. The hope is we can have a weekly "docker --pull" job which will refresh everything once a week, and then all images will use the same base (updating the contents of course). Whilst that may not be in place yet the correct thing for this PR is to assume that fedora:latest is correct and just use it.

FWIW, I don't think this is the correct approach.  Our docker registry is not a "mirroring" process but a "caching" process.  It doesn't _periodically_ refresh what it has, it refreshes what it has based on requests from clients and the age of what it's caching.  If an image goes out of use, it automatically stops getting refreshed as it's never asked for.

So currently, we have these (unnecessary) jobs running on the docker hosts, refreshing their caches, unnecessarily.  It is refreshing caches of a whilelist of images.  That's a whitelist that needs ongoing maintenance as we add more/different docker images.  That refresh is causing pulls to be done for things that might not necessarily be getting used all that often,  and I suspect that problem will only continue to grow as the whitelist will likely never be pruned of images that we stop using.

So ultimately with this method the rate of hits we make to dockerhub will continue to grow and grow (unbounded) as we artificially refresh both the docker host caches and the Nexus registry cache of images we are no longer interested in (because they are not actually in use other than the artificial cache refreshing).  One day this will become a **fire** that needs putting out because we will be refreshing so many unused images that we max out our Docker Hub quota pulling images no longer in use.

Contrast that to my proposal.  Instead of artificially refreshing caches, we let the components function as they are intended to.  That is, every `docker build` includes `--pull`.  That forces the docker hosts to see if they need to pull a new image from Nexus and pull it if necessary -- on every `docker build` attempt.  This keeps the docker host caches fresh automatically, without whitelists that need maintenance, including pruning.  All of those pulls will be satisfied by Nexus without hitting DockerHub for as long as Nexus' cache for that item is considered fresh.  That can be daily, weekly, whatever we want.

On every single one of those pulls (and only on pulls for jobs that actually want the image), when that expiry time for an item expires, Nexus will see if it needs to pull from Docker Hub to refresh it's own cache.  If it does, it passes that back through the docker hosts, refreshing their caches and to the client, to build a new image.

Basically Nexus becomes our rate-limiter for DockerHub while allowing us to hit is as many times as we want for "fresh" images.

In this way, the number of pulls we do from DockerHub is limited to only the number of images currently being used at a frequency of whatever their cache expiry time is in Nexus.  That number (currently_used_images x frequency) becomes the maximum number of hits we make to Docker Hub instead of the alternative above (number_of_images_in_unpruned_whitelist x frequency) where we are making a hit for everything that is in that (potentially, but probably, likely) unpruned whitelist at the same frequency.

Reply:
I would suggest, if we are going to change how we do something at a higher level in how we manage images for reasons that exist the same both before and after this change, that we don't hold up existing PRs to have this debate and let them proceed, and such a debate/change should be its own isolated project.

Reply:
I'm not sure I entirely parsed all of that correctly, but I didn't think we were "holding anything up" here.  This came across as an general enhancement that didn't have any kind of deadline to landing.  The ticket for this says it's targeted for 2.0.

But certainly, if I am mistaken about that, this indeed could land as is, as it ticks the boxes of ensuring the image is always up-to-date as it is, and the refinement of how we fetch images and keep them up-to-date can happen after landing this.

Reply:
It's not urgent in the sense of a release requirement, but, I think, has a lot of benefit we could really use to start benefiting the team immediately, and then perhaps work on such things as follow-ons, if we want to do them.